content/blog/2019-10-23-global-value-numbering/index.md

```diff
@@ -0,0 +1,357 @@
+
+```
+void main(a0 : int, b0 : int, c0 : int, d0 : int, e0 : int, f0 : int) {
+B1:
+  u0 : int = add a0 b0;
+  v0 : int = add c0 d0;
+  w0 : int = add e0 f0;
+  cond : bool = const true;
+  br cond B2 B3;
+B2:
+  x0 : int = add c0 d0;
+  y0 : int = add c0 d0;
+  jmp B4;
+B3:
+  u1 : int = add a0 b0;
+  x1 : int = add e0 f0;
+  y1 : int = add e0 f0;
+  jmp B4;
+B4:
+  u2 : int = phi u0 u1 B2 B3;
+  x2 : int = phi x0 x1 B2 B3;
+  y2 : int = phi y0 y1 B2 B3;
+  z0 : int = add u2 y2;
+  u3 : int = add a0 b0;
+}
+```
+
+Our implementation removes the redundant additions in blocks `B2` and `B3`.
+It also removes from block `B4`: 1) the meaningless phi node `u2 : int = phi u0 u1 B2 B3;`, which has arguments that are always equal, and 2) the redundant phi node `y2 : int = phi y0 y1 B2 B3;`, which is identical to the phi node that precedes it. 
+Here is the full output, which matches the correct output in the paper:
+
+```
+void main(a0 : int, b0 : int, c0 : int, d0 : int, e0 : int, f0 : int) {
+B1:
+  u0: int = add a0 b0;
+  v0: int = add c0 d0;
+  w0: int = add e0 f0;
+  cond: bool = const true;
+  br cond B2 B3;
+B2:
+  jmp B4;
+B3:
+  jmp B4;
+B4:
+  x2: int = phi v0 w0 B2 B3;
+  z0: int = add u0 x2;
+  ret ;
+}
+```
+
+### Trying to get fewer instructions
+
+We treat the number of instructions as a proxy for code performance, where fewer lines of code is superior.
+For each of a number of Bril test programs, we report in the following graph:
+1. The number of instructions after converting to SSA, converting out of SSA, and running Bril's existing [trivial dead code elimination][] (TDCE), shown in blue.
+2. The number of instructions after converting to SSA, running GVN, converting out of SSA, and running TDCE, shown in red.
+
+<img src="eval-correctness.png" width="500"/>
+
+### Bigger benchmarks with TypeScript frontend and comparison with LVN
+
+For a slightly more realistic analysis, we ran GVN on the more complex programs afforded by Bril's TypeScript frontend. In particular, none of these programs were written with intentional redundancy like some of our correctness test cases (thought the conversion from TypeScript to Bril does cause some), so they portray a more realistic evaluation of the benefits of value numbering.
+In addition to the existing test programs from project 1, we also implemented a quadratic formula calculation (`test/gvn/quadratic.ts.bril`), [fizz buzz][] (`test/gvn/fizz-buzz.ts.bril`), and a naive [Sieve of Eratosthenes][] without arrays (`test/gvn/check-primes.ts.bril`).
+All TypeScript programs were first processed by the `ts2bril` frontend.
+We encountered an implementation difficulty: the code generated by the TypeScript frontend sometimes inserts `jmp` instructions after returns in the same basic block.
+We had to manually remove these in order for SSA conversion and dominator tree creation to work.
```

Comment:
Hmm, yeah… I guess maybe the thing to do in the future would be to have the CFG generator transform `ret`s into jumps to a common exit block that just contains `ret` all by itself, in the case where there are multiple return points.